### ISSUE TITLE
Wildcard (*) usage on syscheck

### ISSUE THREAD

**Issue by LegitMigue** [wazuh/wazuh#2723]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.8.2 | Manager/Agent | Any | Unix/Windows |

---

Hello team,

Even if there is no documentation about wildcard usage on Syscheck configuration, wildcards (*) can be used, nevertheless they are not working properly.

**Unix:** Using wildcards (*) on Unix agents will expand the directories/subdirectories search as expected but any new directory that is created on realtime will requiere the agent to be restarted.

Examples using (*): 
`<directories check_all="yes">/opt/test/*</directories>`
`<directories check_all="yes">/opt/test/*/testSub/</directories>`

**Windows:** Using wildcards (*) on Windows is not working at all.

Examples using (*): 
`<directories check_all="yes">E:\test\*</directories>`
`<directories check_all="yes">E:\test\*\testSub\</directories>`

It will be nice for Syscheck to fully support wildcards. Proper documentation will be needed as well.

Cheers!

**Comment by Justangatang** [comment]
We were trying to use wildcards on Windows Syscheck against 'c:\users\*\Downloads' and 'c:\users\*\Desktop'.

That didn't work, and so we had to use the c:\users\ with a restrict="downloads|desktop". This is a little more costly on the system because it has to scan through all the other folders in the c:\users\{usernames} directories. It was also picking up things like desktop.ini and other files with the strings in random folders..

**Comment by jlruizmlg** [comment]
Any update on this issue?

Thanks!

**Comment by rakeshnarayanan1993** [comment]
Hi,

Did you find any update on using wildcards (*) on Windows??

We are trying to collect IIS logs from multiple W3SVC directories
TIA

**Comment by jesusjimsa** [comment]
Hello, @rakeshnarayanan1993.

Sorry for the very late response. Unfortunately, we don't have a fix for this yet. We will do this as soon as we can. In the meantime, you can check the progress in this issue.

**Comment by Molter73** [comment]
We should split this issue in order to better approach how we solve it.

I've opened issue #8259 where we will focus on having FIM (syscheck) reload wildcards while it's running, removing the need to restart Wazuh whenever a directory is added.

We will focus this issue in adding wildcard expansion to Windows. I've been researching this subject and it looks like we might be able to expand wildcards by calling [FindFirstFile](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfilea) and [FindNextFile](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findnextfilea), but some testing is in order before we start to develop the solution using this approach.

**Comment by Molter73** [comment]
After doing some basic testing and re-reading the documentation on `FindFirstFile` and `FindNextFile` I have come to the conclusion that these functions only expand wildcards that are on the "file" part of the path, which means that `C:\testdir\*` is properly expanded to all directories and files contained in `C:\testdir\` but `C:\testdir\*\test` results in `FindFirstFile` failing with error code 123 (`ERROR_INVALID_NAME`).

At this point we have 2 options:
- We can change the wildcards to equivalent regex patterns and go around calling `opendir` and matching each entry with a regex made of partial parts of the configured directory.
- We can wrap calls to `FindFirstFile` and `FindNextFile` in order to split the configured paths, leaving the `*` and `?` always in the "file" part of the path before the calls, we can then recreate the paths recursively until we get the full paths that would match with the wildcards.

Both options look a little tricky, however since the second option would handle opening the directories and does not require us to modify the configured path, I'm leaning more to choosing it.

**Comment by dariommr** [comment]
Hello Team,
I think a global approach to this, could be allowing full regex (OS_Regex), and that should allow us to monitor files in dynamic locations and even certain file types.
For Instance, if we want to monitor all executable files in `%USERPROFILE%\Documents` folder, we could use this regex pattern to match them:
```
  <syscheck>
      <directories check_all="yes" whodata="yes" restrict="\.+\\Documents\\\.+.exe$">C:\Users</directories>
    </syscheck>
```
If we check this pattern against the `ossec-regex` it matches:
```
# /var/ossec/bin/ossec-regex '\.+\\Documents\\\.+.exe$'
Administrator\Documents\test.exe
+OSRegex_Execute: Administrator\Documents\test.exe
+OS_Regex       : Administrator\Documents\test.exe
```
And this should work also in Linux.
I hope this could be helpful.

**Comment by Molter73** [comment]
Hi @dariommr,

That sounds like a good idea. However, when running a scan the pattern specified in the `restrict` attribute is applied to each file we encounter, meaning that it could be executed thousands (maybe even millions?) of times and could potentially result on the scan taking longer. It is also applied to each file when processing a realtime or whodata event, which could also add up on systems with a lot of changing files. We would need to run some tests in order to evaluate if the performance penalty is worth the change.

Could you open a separate issue so we can look into your suggestion at a later time?
